MarkdownFile.js

```diff
@@ -404,11 +403,128 @@ MarkdownFile.prototype._localizeAttributes = function(tagName, tag, locale, tran
     return ret;
 }
 
-var reTagName = /<(\/?)\s*(\w+)(\s|>)/;
-var reSelfClosingTag = /<\s*(\w+)\/>/;
+var reTagName = /^<(\/?)\s*(\w+)(\s+((\w+)(\s*=\s*('((\\'|[^'])*)'|"((\\"|[^"])*)"))?)*)*(\/?)>$/;
+var reSelfClosingTag = /<\s*(\w+)\/>$/;
 var reL10NComment = /<!--\s*[iI]18[Nn]\s*(.*)\s*-->/;
 
-var reDirectiveComment = /<!--\s*i18n-(en|dis)able\s+(\S*)\s*-->/;
+var reDirectiveComment = /i18n-(en|dis)able\s+(\S*)/;
+
+/**
+ * @private
+ * Walk the results of an HTML parse tree to convert it to the
+ * markdown style of nodes.
+ * @param {Node} node the current node of an abstract syntax tree to
+ * walk.
+ * @returns {Array.<Node>} an array of markdown nodes equivalent to
+ * the given HTML node.
+ */
+MarkdownFile.prototype._walkHtml = function(astNode) {
+    var nodes, i, trimmed;
+
+    // for nodes that came from the rehype html parser, convert
+    // them into one or more nodes that remark can process
+    switch (astNode.type) {
+        case 'comment':
+            return new Node({
+                type: "html",
+                value: "<!--" + astNode.value + '-->'
+            });
+
+        case "element":
+            nodes = [];
+
+            // unroll the elements, as that is what markdown does
+            var attrs = [];
+            if (astNode.properties) {
+                for (var name in astNode.properties) {
+                    // for some odd reason, the "class" attribute is renamed to className by the html parser
+                    // so we have to put it back again here
+                    attrs.push((name === "className" ? "class" : name) + '="' + astNode.properties[name] + '"');
+                }
+            }
+            nodes.push(new Node({
+                type: "html",
+                name: astNode.tagName,
+                value: "<" + astNode.tagName + (attrs.length ? " " + attrs.join(" ") : "") + ">"
+            }));
+            if (astNode.children) {
+                astNode.children.forEach(function(child) {
+                    nodes = nodes.concat(this._walkHtml(child));
+                }.bind(this));
+            }
+            nodes.push(new Node({
+                type: "html",
+                name: astNode.tagName,
+                value: "</" + astNode.tagName + ">"
+            }));
+            return nodes;
+
+        case "html":
+            trimmed = astNode.value.trim();
+            // ignore comment html
+            if (trimmed.substring(0, 4) === '<!--') {
+                break;
+            }
+
+            reTagName.lastIndex = 0;
+            match = reTagName.exec(trimmed);
+
+            if (!match) {
+                // this is flow HTML that needs to be parsed into multiple nodes
+                var root = htmlparser.parse(astNode.value);
```

Comment:
Extract parsing nodes from HTML to a separate function (L474-478, L486-505) to a separate function.

Reply:
I don't like to make separate functions that are only called from one spot and are too specific to be called from any other spots. But, I did remove the similar code from lines 746-759 (or so) and instead called the `_walkHtml` method to do the work of converting the node value into mdast nodes in the children, so that code only exists in one spot now.

Reply:
Great!